src/algorithms/synthesis/dd_synthesis.cpp

```diff
@@ -52,4 +54,387 @@ namespace syrec {
+                qcSingleOp.x(currentQubit, ctrlFinal);
+                numGates += 1;
+
+                //return shiftUniquePathsSrc;
+            }
+        }
+        return shiftUniquePathsSrc;
+    }
+
+    auto DDSynthesizer::shiftingPaths(dd::mEdge const& src, dd::mEdge const& current) -> dd::mEdge {
+        dd::mEdge swapPathsSrc;
+        dd::mEdge shiftUniquePathsSrc;
+
+        auto rcSize = qc.size();
+
+        swapPathsSrc = swapPaths(src, current);
+
+        if (qc.size() == rcSize) {
+            rcSize              = qc.size();
+            shiftUniquePathsSrc = shiftUniquePaths(src, current);
+
+        } else {
+            return swapPathsSrc;
+        }
+
+        if (qc.size() == rcSize) {
+            if (terminate(current)) {
+                return src;
+            }
+
+            TruthTable::Cube::Vector p1SigVec;
+            TruthTable::Cube         p1TempVec;
+
+            TruthTable::Cube::Vector p2SigVec;
+            TruthTable::Cube         p2TempVec;
+
+            std::vector<std::size_t> indices;
+
+            pathSignature(current.p->e[0], p1SigVec, p1TempVec);
+            pathSignature(current.p->e[1], p2SigVec, p2TempVec);
+
+            for (std::size_t index = 0; index < p2SigVec.size(); index++) {
+                auto it = std::find(p1SigVec.begin(), p1SigVec.end(), p2SigVec[index]);
+                if (it != p1SigVec.end()) {
+                    indices.emplace_back(index);
+                }
+            }
+
+            return unifyPath(src, current, p1SigVec, p2SigVec, indices);
+        }
+        return shiftUniquePathsSrc;
+    }
+
+    auto DDSynthesizer::synthesizeRec(dd::mEdge const& src, std::unique_ptr<dd::Package<>>& dd) -> dd::mEdge {
+        auto start = std::chrono::steady_clock::now();
+
+        qcSingleOp.clear();
+
+        if (src.p->ref == 0U) {
+            dd->incRef(src);
+        }
+
+        std::queue<dd::mEdge>  q;
+        std::vector<dd::mEdge> visited;
+
+        if (src.p->isIdentity()) {
+            time = static_cast<double>((std::chrono::steady_clock::now() - start).count());
+            return src;
+        }
+
+        q.emplace(src);
+        visited.emplace_back(src);
+
+        while (!q.empty()) {
+            auto newEdge = q.front();
+            q.pop();
+
+            auto oldSize          = qc.size();
+            auto shiftingPathsSrc = shiftingPaths(src, newEdge);
+
+            if (qc.size() > oldSize) {
+                //std::cout << qc.size()<< std::endl;
+
+                auto srcSaved = shiftingPathsSrc;
+
+                auto op          = dd::buildFunctionality(&qcSingleOp, dd);
+                shiftingPathsSrc = dd->multiply(shiftingPathsSrc, op);
+
+                if (shiftingPathsSrc.p->ref == 0U) {
+                    dd->incRef(shiftingPathsSrc);
+                    dd->decRef(srcSaved);
+                    dd->decRef(op);
+                    dd->garbageCollect();
+                }
```

Comment:
These statements should not be placed in an if! As it is right now this would only increase the ref count of new nodes but not of existing ones. This might lead to all kind of problems when garbage collection kicks in.